[Messenger] Resend failed messages to failure transport using a delay from retry strategy

tunterreitmeier opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #53216 #51815
| License       | MIT

This PR tries to fix #53216 and is a follow-up to #51815 
It is somewhat of a stretch to call this a fix, as it changes behavior ~~and especially breaks the BC promise~~.
My sole argumentation for a fix would be that discarding instead of retrying messages is unexpected behavior, but this is subjective.
With changed behavior, I am very much open for comments and suggestions.

**The changes are:**
- `SendFailedMessageToFailureTransportListener` does no longer discard messages if a message was already sent to a failure transport but uses a retry strategy to re-send it to the failure transport using an increasing delay
    - Messages will be retried indefinitely until successfully handled
- `SendFailedMessageToFailureTransportListener` uses the last `ReceivedStamp` in order to not lose information about the original transport
    - It is assumed that the `FailedMessageProcessingMiddleware` is used
    - Otherwise e.g. a handler that is only registered for a specific transport (`from_transport`) would not be considered as a handler when retrieving retries from a failure transport
- The default `MutliplierRetryStrategy` counts all `RedeliveryStamps` to determine the delay instead of the current retry loop
   - Reasons:
       - My argument would be that the delay should increase with every subsequent retrieval from the failure queue, regardless of the current retry in the retry strategy 
       - A retry strategy  with 0 retries would always use the default without increasing at all
- The `AmqpSender` does no longer default to where a message was previously received from as a routing key, when sending it to the failure transport
   - This did not really matter with the default `fanout` binding setup, but interferes with routing in the delay exchange

**Limitations:**
- counting `RedeliveryStamps` is not reliable as `SendFailedMessageForRetryListener` limits (removes) them
    - this could be solved by including a total instead of just the current in the `RedeliveryStamp`

Changing the retry behavior ~~and the public constructor~~ of the EventListener is definitely a breaking change. It's impossible to tell if a project relies on the fact that messages are discarded from the failure transport.
The constructor change is handled gracefully but it still requires some sort of default delay in order to not introduce the regression of an instant and infite retry loop.

Again, really happy for feedback!

tunterreitmeier committed:
[FrameworkBundle] Inject default retry strategy locator into messenger failure listener

tunterreitmeier committed:
[Messenger][Amqp] Do not use redelivery routing key when sending to failure transport

The failure transport uses a delay - the retry routing key from the previous stamp would interfere with publishing to the failure exchange/queue

tunterreitmeier commented on src/Symfony/Component/Messenger/Bridge/Amqp/Tests/Transport/FailureTransportIntegrationTest.php:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @group integration
+ */
+class FailureTransportIntegrationTest extends TestCase
```
Somewhat torn on this test as it requires so much setup.
I created it, as the "failure" logic broke with the change in `AmqpSender` and no auomated test caught it

nicolas-grekas commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -27,11 +29,13 @@
     private ?LoggerInterface $logger;
 
-    public function __construct(ContainerInterface $failureSenders, ?LoggerInterface $logger = null)
+    public function __construct(ContainerInterface $failureSenders, ContainerInterface $retryStrategyLocator, ?LoggerInterface $logger = null)
```
the BC break is not needed and should be avoided
the new argument should be optional and should be added last

tunterreitmeier replied:
I agree and just changed it. My reasoning was that the listener without a retry strategy reintroduces the issue of instant and infinite retries when consuming from the failure transport without some limits. But its not worth breaking  builds over this.

tunterreitmeier committed:
[Messenger] Resend failed messages to failure transport

Use retry strategy with an increasing delay to prevent handling failed messages too fast/often

tunterreitmeier commented:
To me it seems like in these discussions, expecations seem to differ between "I do not want to lose failed messages" and "I do not want an infinite retry loop". So another proposal would be introduce a config flag for a failure transport to either keep or reject failed messages. With 'reject' (or return and ack in reality) being the current behaviour and default. And 're-send to current failure transport without retry strategy / delay' being the new alternative.
But people would have to actively turn on retries in failure transports if they prefer this over losing messages.
With the added side-note that an infinite retry loop is only relevant for `messenger:consume`, not `messenger:failed:retry`. But at least it is fair to assume that you know about this caveat if you actively turned this on.
Again, happy for feedback.

Discussions:
- #53216 
- #51815 
- #63275
